src/node/routes/vscode.ts

```diff
@@ -1,19 +1,30 @@
 import { logger } from "@coder/logger"
 import * as express from "express"
+import * as http from "http"
+import * as net from "net"
 import * as path from "path"
 import { WebsocketRequest } from "../../../typings/pluginapi"
 import { logError } from "../../common/util"
-import { toVsCodeArgs } from "../cli"
+import { CodeArgs, toCodeArgs } from "../cli"
 import { isDevMode } from "../constants"
 import { authenticated, ensureAuthenticated, redirect, self } from "../http"
 import { SocketProxyProvider } from "../socket"
 import { isFile, loadAMDModule } from "../util"
 import { Router as WsRouter } from "../wsRouter"
 import { errorHandler } from "./errors"
 
+export interface IServerAPI {
+  handleRequest(req: http.IncomingMessage, res: http.ServerResponse): Promise<void>
+  handleUpgrade(req: http.IncomingMessage, socket: net.Socket): void
+  handleServerError(err: Error): void
+  dispose(): void
+}
+
+export type CreateServer = (address: string | net.AddressInfo | null, args: CodeArgs) => Promise<IServerAPI>
+
 export class CodeServerRouteWrapper {
   /** Assigned in `ensureCodeServerLoaded` */
-  private _codeServerMain!: CodeServerLib.IServerAPI
+  private _codeServerMain!: IServerAPI
```

Comment:
question: not your code but why do we have an exclamation point in this name?

Reply:
Looks like because we create it on-demand so it might be undefined but we want to use it in a few places without having to keep checking if it is undefined.

IMO this is incorrect usage, we should put it behind a function that checks if it is undefined and throws if so.

Reply:
I think we should make it illegal via eslint actually https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-non-null-assertion.md

Reply:
We could also use `?` instead if it is not important to error.